lucene/core/src/java/org/apache/lucene/index/IndexSorter.java

```diff
@@ -0,0 +1,500 @@
+ */
+public interface IndexSorter {
+
+  /** Used for sorting documents across segments */
+  public interface ComparableProvider {
+    /**
+     * Returns a long so that the natural ordering of long values matches the
+     * ordering of doc IDs for the given comparator
+     */
+    long getAsComparableLong(int docID) throws IOException;
+  }
+
+  /** A comparator of doc IDs, used for sorting documents within a segment */
+  public interface DocComparator {
+    /** Compare docID1 against docID2. The contract for the return value is the
+     *  same as {@link Comparator#compare(Object, Object)}. */
+    int compare(int docID1, int docID2);
+  }
+
+  /**
+   * Get an array of {@link ComparableProvider}, one per segment, for merge sorting documents in different segments
+   * @param readers the readers to be merged
+   */
+  public abstract ComparableProvider[] getComparableProviders(List<? extends LeafReader> readers) throws IOException;
+
+  /**
+   * Get a comparator that determines the sort order of docs within a single Reader.
+   *
+   * NB We cannot simply use the {@link FieldComparator} API because it requires docIDs to be sent
+   * in-order. The default implementations allocate array[maxDoc] to hold native values for comparison,
+   * but 1) they are transient (only alive while sorting this one segment) and 2) in the typical
+   * index sorting case, they are only used to sort newly flushed segments, which will be smaller
+   * than merged segments
+   *
+   * @param reader the Reader to sort
+   */
+  public abstract DocComparator getDocComparator(LeafReader reader) throws IOException;
+
+  /**
+   * Serializes the parent SortField.  This is used to write Sort information into the Segment header
+   *
+   * @see SortFieldProvider#loadSortField(DataInput)
+   */
+  public abstract void serialize(DataOutput out) throws IOException;
+
+  /**
+   * The SPI-registered name of a {@link SortFieldProvider} that will deserialize the parent SortField
+   */
+  public abstract String getProviderName();
+
+  /**
+   * Provide a NumericDocValues instance for a LeafReader
+   */
+  public interface NumericDocValuesProvider {
+    /**
+     * Returns the NumericDocValues instance for this LeafReader
+     */
+    NumericDocValues get(LeafReader reader) throws IOException;
+  }
+
+  /**
+   * Provide a SortedDocValues instance for a LeafReader
+   */
+  public interface SortedDocValuesProvider {
+    /**
+     * Returns the SortedDocValues instance for this LeafReader
+     */
+    SortedDocValues get(LeafReader reader) throws IOException;
+  }
+
+  /**
+   * Serialize an object into a DataOutput
+   */
+  public interface Serializer {
+    /** Serializes an object into a DataOutput */
+    void serialize(DataOutput out) throws IOException;
+  }
+
+  /**
+   * Sorts documents based on integer values from a NumericDocValues instance
+   */
+  public static final class IntSorter implements IndexSorter {
```

Comment:
can it be package private?

Reply:
IndexSorter is an interface (started as an abstract class in a first iteration) so everything is public and static, and we don't need those modifiers anymore.